### ISSUE TITLE
Update sync fails with "A SqlParameter with ParameterName '@changeTable' is not contained by this SqlParameterCollection

### ISSUE THREAD

**Issue by DeakinWilson** [Mimetis/Dotmim.Sync#356]
I'm not sure if this is a bug, or I just need some help...

I have a SQL Server server and a Sqlite client (on iOS).  I'm using the WebProxy method for getting and retrieving data.  I'm also using Filters.

I had some trouble with the filters initially so tried to re-provision the server.  That gave a number of errors so I ended up manually removing all the stored procedures (these steps are possibly my downfall).

The initial sync works great, lots of Stored Procedures are (re-)created on the SQL Server and data is sync'd to Sqlite.  All is working as expected.  I can view the data in the local sqlite database.

I then add a row to the Sqlite client and perform another sync.  The progress tracking received is:

`BeginSession : 8:12:38 PM.799	 Session Id:2623dcc8-3b2c-4bb1-9cfe-e457340adb33`
`ScopeLoaded : 8:12:39 PM.36	 [main] [DefaultScope] [Version 1] `
`Last sync:10/27/2020 6:56:56 AM Last sync duration:0:0:0.662 `
`ChangesSelecting : 8:12:40 PM.973	 [main] [Tickets] upserts:1 deletes:0 total:1 `
`ChangesSelected : 8:12:40 PM.993	 [main] upserts:1 deletes:0 total:1 `

The error received is:

<code>
SyncDataAsync: Failed to synchronize data : Dotmim.Sync.SyncException: A SqlParameter with ParameterName '@changeTable' is not contained by this SqlParameterCollection.
  at Dotmim.Sync.Web.Client.HttpRequestHandler.HandleSyncError (System.Net.Http.HttpResponseMessage response, Dotmim.Sync.Serialization.ISerializerFactory serializerFactory) [0x0022f] in <b7eccb9974e74dd193f6a7e66f943841>:0 
  at Dotmim.Sync.Web.Client.HttpRequestHandler.ProcessRequestAsync[U] (System.Net.Http.HttpClient client, System.String baseUri, System.Byte[] data, Dotmim.Sync.Web.Client.HttpStep step, System.Guid sessionId, System.String scopeName, Dotmim.Sync.Serialization.ISerializerFactory serializerFactory, Dotmim.Sync.Serialization.IConverter converter, System.Int32 batchSize, System.Threading.CancellationToken cancellationToken) [0x0060b] in <b7eccb9974e74dd193f6a7e66f943841>:0 
  at Dotmim.Sync.Web.Client.WebClientOrchestrator.ApplyThenGetChangesAsync (Dotmim.Sync.ScopeInfo scope, Dotmim.Sync.Batch.BatchInfo clientBatchInfo, System.Threading.CancellationToken cancellationToken, System.IProgress`1[T] progress) [0x003aa] in <b7eccb9974e74dd193f6a7e66f943841>:0 
  at Dotmim.Sync.SyncAgent.SynchronizeAsync (Dotmim.Sync.Enumerations.SyncType syncType, System.Threading.CancellationToken cancellationToken, System.IProgress`1[T] progress) [0x00d1c].
</code>

**Comment by Mimetis** [comment]
This exception happens when you are missing a stored procedure.
If you are in a test mode, just start with a fresh blank database, where nothing has been provisioned yet.

If you are not in a test mode, you may want to DeprovisionAsync and ProvisionAsync again to restore all the stored procedure

**Comment by DeakinWilson** [comment]
Thanks for the reply.  I did have a question about DeprovisionAsync and ProvisionAsync - I'm running in WebProxy mode and so wonder if I need to run those methods from the Server, or if I can do it from the Client?

**Comment by Mimetis** [comment]
depends if you need to deprovision both 
If you need only server, create a console app with a RemoteOrchestrator and call Provision / Deprovision
If you need only client, make a LocalOrchestrator instead

**Comment by DeakinWilson** [comment]
Thank you.  I've now run the DeprovisionAsync for the server and get errors for the missing stored procedures.  I'd like to suggest a modification to the "DELETE PROC" sql statements to include "IF EXISTS" before the delete.  

I have solved the issue by creating dummy procedures for the missing ones for now and, with the ProvisionAsync, have solved the original issue.

Thanks again.

**Comment by REMjn832** [comment]
I am having this same problem all of the sudden. I am using the 'SqlSyncChangeTrackingProvider' with 2 SQL-Server databases that I thought were working (synced fine when there was no data). I re-provisioned them (de-provision and re-provision) which worked. However, as soon as I add data to the server that should be synced to the client, this error will be thrown, and the sync fails. I can re-provision successfully still, and if I remove that row from the server, the sync succeeds again.

> "Dotmim.Sync.SyncException: A SqlParameter with ParameterName '@changeTable' is not contained by this SqlParameterCollection.\r\n ---> System.IndexOutOfRangeException: A SqlParameter with ParameterName '@changeTable' is not contained by this SqlParameterCollection.\r\n   at Microsoft.Data.SqlClient.SqlParameterCollection.GetParameter(String parameterName)\r\n   at Dotmim.Sync.SqlServer.Builders.SqlSyncAdapter.ExecuteBatchCommandAsync(DbCommand cmd, Guid senderScopeId, IEnumerable`1 applyRows, SyncTable schemaChangesTable, SyncTable failedRows, Int64 lastTimestamp, DbConnection connection, DbTransaction transaction)\r\n   at Dotmim.Sync.SyncAdapter.ApplyBulkChangesAsync(Guid localScopeId, Guid senderScopeId, SyncTable changesTable, Int64 lastTimestamp, List`1 conflicts, DbConnection connection, DbTransaction transaction)\r\n   at Dotmim.Sync.CoreProvider.ApplyChangesInternalAsync(SyncTable schemaTable, SyncContext context, MessageApplyChanges message, DbConnection connection, DbTransaction transaction, DataRowState applyType, DatabaseChangesApplied changesApplied, CancellationToken cancellationToken, IProgress`1 progress)\r\n   at Dotmim.Sync.CoreProvider.ApplyChangesAsync(SyncContext context, MessageApplyChanges message, DbConnection connection, DbTransaction transaction, CancellationToken cancellationToken, IProgress`1 progress)\r\n   at Dotmim.Sync.LocalOrchestrator.ApplyChangesAsync(ScopeInfo scope, SyncSet schema, BatchInfo serverBatchInfo, Int64 clientTimestamp, Int64 remoteClientTimestamp, ConflictResolutionPolicy policy, CancellationToken cancellationToken, IProgress`1 progress)\r\n   --- End of inner exception stack trace ---\r\n   at Dotmim.Sync.BaseOrchestrator.RaiseError(Exception exception)\r\n   at Dotmim.Sync.LocalOrchestrator.ApplyChangesAsync(ScopeInfo scope, SyncSet schema, BatchInfo serverBatchInfo, Int64 clientTimestamp, Int64 remoteClientTimestamp, ConflictResolutionPolicy policy, CancellationToken cancellationToken, IProgress`1 progress)\r\n   at Dotmim.Sync.SyncAgent.SynchronizeAsync(SyncType syncType, CancellationToken cancellationToken, IProgress`1 progress)\r\n   at OTC.DataAccess.CustomerPortal.SynchronizeDatabasesAsync()...

**Comment by Mimetis** [comment]
You miss a stored procedure somewhere.
It's quite complicated to debug here.

On try you can do is testing in a new clean environment to see if you are able to reproduce easily.
Once it's done, you may want to share the repro with me

**Comment by REMjn832** [comment]
Thanks. 
I did try on a fresh database, and it worked fine. If I am missing a stored procedure, how do I know which one is missing? At one point, I had to start over, because re-provisioning was not applying a change to a column that I made (not a PK column, just a non-indexed column from nvarchar to bigint). Since it would not apply the changes from the server to the client, I removed all triggers, and stored procedures and scope tables (from both server and client databases). Then I tried to sync again, and it worked fine - until I added the first row of data on the server, and then the error above was introduced. I am not sure what to do. Thanks.
